src/Symfony/Component/DependencyInjection/Loader/XmlFileLoader.php

```diff
@@ -505,12 +506,13 @@ private function getArgumentsAsPhp(\DOMElement $node, $name, $file, $lowercase =
                 case 'collection':
                     $arguments[$key] = $this->getArgumentsAsPhp($arg, $name, $file, false);
                     break;
+                case 'service_locator':
                 case 'iterator':
                     $arg = $this->getArgumentsAsPhp($arg, $name, $file, false);
                     try {
-                        $arguments[$key] = new IteratorArgument($arg);
+                        $arguments[$key] = 'iterator' === $type ? new IteratorArgument($arg) : new ServiceLocatorArgument($arg);
```

Comment:
given that all the parsing of referenced is already factored inside `getArgumentsAsPhp`, I don't think it makes sense to group the 2 cases together and then performed `if` checks in all other statements